core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -438,18 +438,33 @@ public boolean hasAliases(final String[] aliases, String[] concreteIndices) {
 
     /**
      * Returns indexing routing for the given index.
+     * If routing is not null then the value is returned, parent routing and alias routing are ignored.
+     * If routing is null and there is an alias routing then the value is returned and parent is ignored.
+     * If routing is null and a parent is defined then the parent is returned (unless an alias routing is defined, see above).
      */
     // TODO: This can be moved to IndexNameExpressionResolver too, but this means that we will support wildcards and other expressions
     // in the index,bulk,update and delete apis.
-    public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex) {
-        if (aliasOrIndex == null) {
+    public String resolveIndexRouting(@Nullable String parent, @Nullable String routing, String aliasOrIndex) {
+        if (routing != null) {
+            // The routing is explicit, ignore the alias and parent routing (if any).
```

Comment:
I think the main use-case for routed aliases is to be able to transparently query a multi-tenant index holding data for several users as if each user had a dedicated index. The consumers of these aliases might not even know that these are aliases, it might be something that is set up by the system administrator. So I think we need the following rules when an alias routing is configured:
 - alias _routing wins over _parent (like the PR already does)
 - if _routing is set and there is an alias _routing, fail if the routing keys are different (like before) otherwise we would either allow to put documents in shards that are not visible to the alias (bad) or silently ignore the user-provided routing key which I believe could be confusing. Given that routing is an expert feature, I think it's totally fine to fail in that case: if an alias with routing is configured, this is exactly so that consumers of this alias don't have to deal with routing explicitly?